universe/ValueRef.h

```diff
@@ -83,5 +90,87 @@ enum StatisticType : int {
 
 }
 
+//! Holds all FreeOrion named ValueRef%s.  ValueRef%s may be looked up by name.
+class NamedValueRefManager {
+public:
+    //using container_type = std::map<const std::string, const std::unique_ptr<ValueRef::ValueRefBase>>;
+    using key_type = std::string;
+    using value_type = std::unique_ptr<ValueRef::ValueRefBase>;
+    using container_type = std::map<key_type, value_type>;
+
+    using iterator = container_type::const_iterator;
+
+    //! Returns the ValueRef with the name @p name or nullptr if there is nov ValueRef with such a name or of the wrong type
+    //! use the free function GetValueRef(...) instead, mainly to save some typing.
+    template <typename T>
+    auto GetValueRef(const std::string& name) -> ValueRef::ValueRef<T>*;
+
+    //! Returns the ValueRef with the name @p name; you should use the
+    //! free function GetValueRef(...) instead, mainly to save some typing.
+    auto GetValueRefBase(const std::string& name) const -> ValueRef::ValueRefBase*;
+
+    auto NumNamedValueRefs() const -> std::size_t { return m_value_refs.size(); }
+
+    //! iterator to the first value ref
+    FO_COMMON_API auto begin() const -> iterator;
+
+    //! iterator to the last + 1th value ref
+    FO_COMMON_API auto end() const -> iterator;
+
+    // Singleton
+    NamedValueRefManager&  operator=(NamedValueRefManager const&) =delete; // no copy via assignment
+
+    NamedValueRefManager(NamedValueRefManager const&) =delete;            // no copies via construction
+
+    ~NamedValueRefManager();
+
+    /** Sets named value refs to the value of \p future. */
+    FO_COMMON_API void SetNamedValueRefs(Pending::Pending<container_type>&& future);
+
+    //! Returns the instance of this singleton class; you should use the free
+    //! function GetNamedValueRefManager() instead
+    static NamedValueRefManager& GetNamedValueRefManager();
+
+    //! Returns a number, calculated from the contained data, which should be
+    //! different for different contained data, and must be the same for
+    //! the same contained data, and must be the same on different platforms
+    //! and executions of the program and the function. Useful to verify that
+    //! the parsed content is consistent without sending it all between
+    //! clients and server.
+    auto GetCheckSum() const -> unsigned int;
+
+    //! Register the @p value_ref under the evaluated @p name.
+    template <typename T>
+    void RegisterValueRef(std::string name, std::unique_ptr<T> vref);
```

Comment:
`std::unique_ptr<ValueRef::ValueRef<T>>` ?

Reply:
i think that would disallow use of RegisterValueRefBase, but probably we do not need that right now

Reply:
done `std::unique_ptr<T>` -> `std::unique_ptr<ValueRef::ValueRef<T>>`